### PR TITLE
Fix a few issues with PipeStream

### LINKED ISSUE #12: PipeStream returns 0 instead of blocking when running a command with intermittent output
For details, you can see [this issue](http://stackoverflow.com/q/37059305/516433) on StackOverflow.

### LINKED ISSUE #142: SshCommand OutputStream has no way to directly set BlockLastReadBuffer
When reading the stdout of a process executed through SshCommand the output stream will throw an exception if you read beyond what's available in the buffer and it seems to be by design.

Although in my opinion this is the wrong default, there's no direct way to change the property at runtime as the exposed interface is the standard `Stream`.

Could the default be changed to more sensible blocking behavior or there be a public method/property that can be used to change it at runtime?

### LINKED ISSUE #164: Cannot get streaming stdout/stderr and wait for remote process to exit
My scenario is that I want to start a process on a remote machine and wait for it to either crash, or exit gracefully. While I am waiting for it to crash/exit I want to propagate it's stdout/stderr to the host machine so I can parse it, or redirect it to file once I get the data I need from it.

My attempt is: https://gist.github.com/alanmcgovern/c12867f3accc6cf5870d5212d3425974#file-test-cs-L8

The issue with this approach is that the very first `(line = reader.ReadLine ())` returns null if you do it 'too quickly'. PipeStream thinks the stream has completed even though the process is still running on the remote machine. If i put a Thread.Sleep before that line I can read some output before i get a null, but eventually i get a null when reading, even though the process hasn't exited.

The only 'reliable' approach to detect when my process exits is to create a ShellStream, then write this command: `path/to/long_lived_application; echo TOMBSTONE;` and then exit my read-loop when i see the string `TOMBSTONE` appear in the output.

This approach is here: https://gist.github.com/alanmcgovern/c7bf0fab9760cfcd324573df51c98fbb

Is there a better way to asynchronously run a command and gather it's stdout/stderr?

### LINKED ISSUE #440: PipeStream not supporting Read offsets > 0
PipeStream seems to throw an exception when Read(byte[] buffer, int offset, int count) is called, with an offset above zero. 
https://github.com/sshnet/SSH.NET/blob/master/src/Renci.SshNet/Common/PipeStream.cs#L183

Is there a reason it cannot support this?  Seems like it would be easy enough to modify the for loop to start at the desired offset, and still calculate the amount read properly.

If it's just something that hasn't been done yet, I can perhaps do the work, but I wanted to check first to see if there's some edge case that makes this feature N/A that I'm not seeing.

Thanks!

### LINKED ISSUE #1045: No response due to the file size over the default MaxBufferLength of PipeStream
The default MaxBufferLength of PipeStream is _maxBufferLength = 200 * 1024 * 1024
The issue occurs when there is a file in linux, whose sizes are over the default MaxBufferLength.
Steps:
1. Given a JsonData1.txt, whose sizes are 210MB ( > 200MB), and a JsonData2.txt, whose sizes are 190MB (< 200MB)
2. Copy both JsonData1.txt and JsonData2.txt  into /test/ in your linux.
3. Test the JsonData by attached test.zip file.
4. Config commandtext= "cat /test/JsonData1.txt" in app.config, and config other items.
5. Run it, then you will get a time out issue. Actually, it should spend about 17s, if you change _maxBufferLength = **300** * 1024 * 1024
6. Config commandtext= "cat /test/JsonData2.txt" in app.config.
7. Run it, then you will get an available timespan.

[Test.zip](https://github.com/sshnet/SSH.NET/files/10011454/Test.zip)

### PR DESCRIPTION
**Pull request by Rob-Hague**
Apply similar treatment to PipeStream as #1322 did to ShellStream

PipeStream now behaves much more Stream-like. In particular, it performs partial reads (instead of blocking until a certain amount of data is available), blocks until data is available (instead of returning 0 prematurely) and removes the Stream-unlike properties `BlockLastReadBuffer` and `MaxBufferLength`.

Sadly I gave up trying to make a benchmark compatible with all the quirks of the previous implementation, but a dumb throughput test (reading and writing simultaneously) shows about 5.2GB/s with this implementation compared to 140MB/s previously.

Some cleanup of its usage in the library followed.

contributes to #650 (does not make any of the suggested source breaking changes - still worth considering)
closes #440
closes #164
closes #12
closes #1045
closes #142

of existing PRs:
closes #1070
closes #924
closes #374
closes #13

### CODE REVIEW

File: test/Renci.SshNet.IntegrationTests/SshTests.cs [hunk 1]
```diff
@@ -297,8 +289,16 @@ public void Ssh_Command_IntermittendOutput_OutputStream()
...
+                        // command.Result (also returned from EndExecute) consumes OutputStream,
+                        // which we've already read from, so Result will be empty.
+                        // TODO consider the suggested changes in https://github.com/sshnet/SSH.NET/issues/650
+
+                        //Assert.AreEqual(expectedResult, actualResult);
+                        //Assert.AreEqual(expectedResult, command.Result);
+
+                        // For now just assert the current behaviour.
+                        Assert.AreEqual(0, actualResult.Length);
+                        Assert.AreEqual(0, command.Result.Length);
```
Comment (Rob-Hague) [inline test/Renci.SshNet.IntegrationTests/SshTests.cs:301]: I was initially planning on getting this PR in, then getting a release out. Now I am thinking of addressing #650 beforehand by taking the break changing `string Execute()` to `void Execute()` in order to stop `Execute` from consuming `OutputStream`.

That way, we can get the CancelAsync changes (#1345), this PR with its fixes and subtle behaviour changes, and the source-breaking changes wrapped up in one release. Probably easier for consumers.

**Review by WojciechNagorski — APPROVED** [review]
LGTM!

**Comment by Rob-Hague** [comment]
Thank you

**Comment by vladd** [comment]
Thank you for fixing all these problems! Is there any estimation, in which release is the fix going to be included?

**Comment by Rob-Hague** [comment]
I have a couple more fixes/changes I'd like to make to SshCommand before the next release, and then we can make one then. So I would say June, but the last time I answered that question I said May and it won't be May :-)

**Pull request merged**